components/filesystem/include/filesystem/filesystem.hpp

```diff
@@ -0,0 +1,84 @@
+/* Copyright (c) 2024, Thomas Atkinson
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 the "License";
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <filesystem>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "core/platform/context.hpp"
+#include "core/util/logging.hpp"
+
+namespace vkb
+{
+namespace filesystem
+{
+struct FileStat
+{
+	bool   is_file;
+	bool   is_directory;
+	size_t size;
+};
+
+using Path = std::filesystem::path;
+
+// A thin filesystem wrapper
+class FileSystem
+{
+  public:
+	FileSystem()          = default;
+	virtual ~FileSystem() = default;
+
+	virtual FileStat             stat_file(const Path &path)                                    = 0;
+	virtual bool                 is_file(const Path &path)                                      = 0;
+	virtual bool                 is_directory(const Path &path)                                 = 0;
+	virtual bool                 exists(const Path &path)                                       = 0;
+	virtual bool                 create_directory(const Path &path)                             = 0;
+	virtual std::vector<uint8_t> read_chunk(const Path &path, size_t start, size_t offset)      = 0;
+	virtual void                 write_file(const Path &path, const std::vector<uint8_t> &data) = 0;
+	virtual void                 remove(const Path &path)                                       = 0;
+
+	virtual const Path &external_storage_directory() const = 0;
+	virtual const Path &temp_directory() const             = 0;
+
+	void write_file(const Path &path, const std::string &data);
+
+	// Read the entire file into a string
+	std::string read_file_string(const Path &path);
+
+	// Read the entire file into a vector of bytes
+	std::vector<uint8_t> read_file_binary(const Path &path);
```

Comment:
```suggestion
	std::vector<uint8_t> read_file_binary(const Path &path);
	void with_file_contents(const Path& path, const std::function<void(const uint8_t*, size_t)>& handler);
```
In my experience, most applications only need the asset data for the scope of the call where it's read. While read_binary_file returns a std::vector and thus always allocates heap space, the implementation of something like with_file_contents is free to give a pointer to a memory mapped file, or something like the result of an AAsset_getBuffer(...) call on Android, which can lower the amount of actual heap space used in the application. Obviously anything going into a staging buffer will consume host memory anyway, but at least this would potentially cut that in half.

Reply:
I tried this on a local branch starting from this PR branch. I updated the GLTF loader code and the image loading code to both use loading from memory mapped files instead of from vectors.  I then ran `time ./build/app/bin/release/AMD64/vulkan_samples.exe --stop-after-frame 100 sample wait_idle` on both my branch and on the main branch.  

However, load performance was not appreciably different between the two branches.  Therefore unless there are actually issues with memory constrained platforms, I'm guessing that the increased complexity of using memory mapping APIs isn't going to be worth it.

I may try some other platforms later, but it looks as if the file IO streams API has significantly improved in performance on Windows (relative to memory mapped files) since the last time I did work like this.

Reply:
Thank you for benchmarking the differences! From what i have found in the past most of our slow loading times comes from decompressing ASTC if a platform doesnt support this format. See #860 as a reference to this topic

Reply:
Resolving this for now but definitely open to adding a memory mapped alternative